Models/book

juanmbellini opened the pull request:
# Summary

This PR adds the Book model, according to the trello card: [https://trello.com/c/eZNQaOU5](https://trello.com/c/eZNQaOU5)

# Notes

- Replace the ddl-auto property to none, and use [flyway](https://flywaydb.org/) for schema generation, using SQL migration scripts.
- Use the ```orm.xml``` file instead of JPA annotations (this makes a better separation of concerns between the models and the persistence metadata).

The base branch is feature/welcome, which should be changed to dev once [the feature/welcome PR](https://github.com/wolox-training/jmb-java/pull/2) is merged.

# Trello Card

https://trello.com/c/eZNQaOU5

juanmbellini committed:
Add the book model

juanmbellini committed:
Add Flyway

- Include the Book model schema creation migration script

juanmbellini committed:
Add JPA mapping

- Use orm.xml for the Object-Relational mapping as this approach does not mix domain objects stuff with persistence stuff (makes a better separation of concerns)

juanmbellini committed:
Add equals, hashcode and toString methods to the Book model

juanmbellini committed:
Change ddl-auto property to none

- Now the schema is generated by Flyway

darkBasic666 commented on src/main/java/wolox/training/models/Book.java:
```diff
@@ -0,0 +1,294 @@
+    private static void assertAuthor(final String author) {
+        Assert.notNull(author, "The author must not be null");
+    }
+
```
Why add this asserts? Is for check null before insert in DB ?

juanmbellini replied:
Yes. I prefer the models to validate this stuff themselves, and leave the upper layer (a.k.a the service layer) to validate business stuff (for example, age of a user, and that sort of stuff).

darkBasic666 commented on src/main/resources/META-INF/orm.xml:
```diff
@@ -0,0 +1,46 @@
+      </basic>
+    </attributes>
+  </entity>
+</entity-mappings>
```
Excelent work.

juanmbellini replied:
thank you

darkBasic666 commented on src/main/java/wolox/training/models/Book.java:
```diff
@@ -0,0 +1,294 @@
+            ", pages=" + pages +
+            ", isbn='" + isbn + '\'' +
+            '}';
+    }
```
Maybe using [Lombok](https://projectlombok.org/features/Data) @Data reduce ball of text of this class.

juanmbellini replied:
Didn't know this project. I'll read about it now

juanmbellini replied:
Okay, replaced stuff with lombok

juanmbellini committed:
Add lombok

- Remove getters, toString, hashcode and equals to the Book class and replace with lombok

juanmbellini committed:
Fix database schema

- Year was being represented with an int when requirements indicate that it is represented with a String
- Pages was being represented with a String when requirements indicate that it is represented with an int.

juanmbellini committed:
Remove the @Data annotation

- It is not needed

juanmbellini committed:
Make the Book default constructor be package private

- This constructor must be only used by the JPA provider
